server/app/services/ProgramBlockValidation.java

```diff
@@ -48,26 +57,28 @@ public enum AddQuestionResult {
    * creating to ensure that no one can maliciously mess up data by sending specially crafted
    * request (or in case we mess up client-side checks).
    */
-  public static AddQuestionResult canAddQuestion(
+  public AddQuestionResult canAddQuestion(
       ProgramDefinition program, BlockDefinition block, QuestionDefinition question) {
+    if (versionRepository
+        .getDraftVersion()
```

Comment:
**issue (blocking):** this is going to query the database for the current draft version each time the method is called, which we do once per question in `ProgramQuestionBank`.

The least disruptive way I can think of to address this is to inject a `ProgramBlockValidationFactory` in the places that currently consume `ProgramBlockValidation`. The factory then creates an instance of `ProgramBlockValidation` which can be created and reused locally to prevent N+1 database queries i.e. the instance can be retrieved before the call to `return params.questions().stream()` and reused inside the call to `filter()`

Reply:
I think I understand what your saying - instead of creating an instance of VersionRepository for each call, creating an instance of ProgramBlockValidation and reusing the same object will be prevent excess DB calls. Am I getting it rt?

Reply:
Also, should the Unit test be updated?